### PR TITLE
batches: add button to automatically exclude repo in batch spec

### PR DESCRIPTION
**Pull request by courier-new**
PoC that adds a button to the workspaces preview list items to auto-update the batch spec code in the editor when a user wants to exclude a repo originally returned from the search results. Uses `yaml-ast-parser` to parse an AST from the batch spec, identify where to append `-repo:` and/or which `repository:` directive to remove, and manipulates the batch spec to exclude the desired repo/workspace.

It's pretty hackily implemented, and it's kinda janky since the workspaces preview UI takes a couple seconds to update and reflect the new workspaces without the one that was excluded, but it works!!

https://user-images.githubusercontent.com/8942601/133354872-fa9b489b-e6b1-4a86-99ae-cca6208d6a57.mov

Left a couple questions in the form of TODOs.

Also for the record, [yaml-ast-parser](https://github.com/mulesoft-labs/yaml-ast-parser) looks to be abandoned, but it also has zero dependencies and seems to work pretty reliably, so I'm not too concerned about using it. But let me know if you feel otherwise.

### CODE REVIEW

File: client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx [hunk 1]
```diff
@@ -177,31 +203,48 @@ const PreviewWorkspaces: React.FunctionComponent<{ preview: BatchSpecWorkspacesF
...
-                                    <span className="text-muted">{step.container}</span>
-                                </li>
-                            ))}
-                        </ul>
+                        <button
+                            className="btn align-self-start p-0 m-0 mr-3"
+                            data-tooltip="Omit this repository from batch spec file"
+                            type="button"
+                            // TODO: Is displayName the one that would match whatever
+                            // branch is specified in the batch spec file?
```
Comment (courier-new) [inline client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx:223]: This seemed to be true from all the examples I tried to test but I want to make sure this is an accurate assumption.

File: client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx [hunk 2]
```diff
@@ -177,31 +203,48 @@ const PreviewWorkspaces: React.FunctionComponent<{ preview: BatchSpecWorkspacesF
...
+                            // TODO: Are there weird exceptions that we can't reliably
+                            // figure out how to exclude and should just disable this
+                            // button for in the short term?
```
Comment (courier-new) [inline client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx:226]: Basically I'm wondering if there's any weird edge cases for how one could specify a workspace in `on` in the spec that we couldn't just remove by the repo/branch names.

File: client/web/src/enterprise/batches/create/yaml-util.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    // Take the value for the "repositoriesMatchingQuery" key -- this should be a
+    // `YAMLScalar` for the query search string
+    const queryValue = queryMapping?.value
+
+    if (!queryValue || !isYAMLScalar(queryValue)) {
+        return { spec, success: false, error: 'Non-scalar value found for "repositoriesMatchingQuery" key' }
+    }
+
+    // Insert "-repo:" qualifier at the end of the query string
+    // TODO: Does this always work if it's at the end of the search query string?
```
Comment (courier-new) [inline client/web/src/enterprise/batches/create/yaml-util.ts:72]: Again, this seemed to work from all my testing, but I wanted to double-check.

**Review by eseliger — APPROVED** [review]
That's awesome! Thanks so much for looking into it and finding a good AST library!

File: client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx (hunk 1 above)
Comment (mrnugget) [reply client/web/src/enterprise/batches/create/examples/ExampleTabs.tsx:223]: I think so. Leaving aside the special casing of pull requests here the `displayName` is the full ref minus `refs/heads`:

https://github.com/sourcegraph/sourcegraph/blob/b3de0faeeba499bd20b5dc61ef96b3977c5d1fee/cmd/frontend/graphqlbackend/git_ref.go#L48-L61

I don't think users specify `refs/heads/foobar` in the batch spec.

But `abbrevName` seems to be the same without the fancy PR things:

https://github.com/sourcegraph/sourcegraph/blob/b3de0faeeba499bd20b5dc61ef96b3977c5d1fee/cmd/frontend/graphqlbackend/git_ref.go#L111-L112

File: client/web/src/enterprise/batches/create/yaml-util.test.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+        it('should fail and return an error if it cannot parse the spec', () => {
+            expect(excludeRepo('invalid', 'repo1', 'doesnt-matter')).toEqual({
+                success: false,
+                error: 'Spec not parseable',
+                spec: 'invalid',
+            })
+        })
+    })
+})
```
Comment (mrnugget) [inline client/web/src/enterprise/batches/create/yaml-util.test.ts:204]: Good stuff!

File: client/web/src/enterprise/batches/create/yaml-util.ts (hunk 3 above)
Comment (mrnugget) [reply client/web/src/enterprise/batches/create/yaml-util.ts:72]: Agree with Erik. The "correct" thing would be to later extend the batch spec schema so we can do stuff like:

```yaml
on:
  - repositoriesMatchingQuery: guess-who type:repo
  - repository: github.com/sourcegraph/hidden-laser
  - exclude:
    - repository: github.com/sourcegraph/archived-thing
```

File: client/web/src/enterprise/batches/create/yaml-util.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            trimLastSequenceItemDelimiter(spec.slice(0, branchMatchMap.startPosition)) +
+            spec.slice(branchMatchMap.endPosition),
+    }
+}
+
+/**
+ * Trims the final sequence delimiter (i.e. a set of newlines, spaces, and a dash) from
+ * the given slice of raw batch spec.
+ *
+ * This is "sorry-pls-don't-hate-me"-level hack but unfortunately the easiest way around a
```
Comment (mrnugget) [inline client/web/src/enterprise/batches/create/yaml-util.ts:173]: Hey, [it's all fair in love and war](https://en.wikipedia.org/wiki/All_Is_Fair_in_Love_and_War) _aaand_ in AST-based transformation of source code.

**Review by mrnugget — APPROVED** [review]
Good stuff! This looks like it was fun :)

**Review by LawnGnome — APPROVED** [review]
I really like the AST approach here, and the code is still nicely readable. Good job!

**Pull request merged**